lib/service/certreloader_test.go

```diff
@@ -0,0 +1,269 @@
+				require.Len(t, after, 2)
+				require.NotEqual(t, before[0], after[0])
+				require.Equal(t, before[1], after[1])
+			},
+		},
+		{
+			desc: "c1 and c2 certs do change with an update",
+			certsUpdate: func(t *testing.T, certs *certFiles) {
+				// Update c1 cert.
+				key, crt := newCertKeyPair(t)
+				write(t, certs.c1Key, key)
+				write(t, certs.c1Crt, crt)
+
+				// Update c2 cert.
+				key, crt = newCertKeyPair(t)
+				write(t, certs.c2Key, key)
+				write(t, certs.c2Crt, crt)
+			},
+			certsReloadErrorAssert: require.NoError,
+			certsAssert: func(t *testing.T, before []tls.Certificate, after []tls.Certificate) {
+				// Both certs have been updated.
+				require.Len(t, before, 2)
+				require.Len(t, after, 2)
+				require.NotEqual(t, before[0], after[0])
+				require.NotEqual(t, before[1], after[1])
+			},
+		},
+		{
+			desc: "c1 and c2 certs do not change with an incomplete update",
+			certsUpdate: func(t *testing.T, certs *certFiles) {
+				// Update c1 cert.
+				key, crt := newCertKeyPair(t)
+				write(t, certs.c1Key, key)
+				write(t, certs.c1Crt, crt)
+
+				// Update only c2 key.
+				key, _ = newCertKeyPair(t)
+				write(t, certs.c2Key, key)
+			},
+			certsReloadErrorAssert: require.Error,
+			certsAssert: func(t *testing.T, before []tls.Certificate, after []tls.Certificate) {
+				// No cert has been updated.
+				require.Len(t, before, 2)
+				require.Len(t, after, 2)
+				require.Equal(t, before[0], after[0])
+				require.Equal(t, before[1], after[1])
+			},
+		},
+		{
+			desc: "c1 cert does not change during an ongoing update",
+			certsUpdate: func(t *testing.T, certs *certFiles) {
+				// Update c1 key, and partially update c1 cert.
+				key, crt := newCertKeyPair(t)
+				write(t, certs.c1Key, key)
+				write(t, certs.c1Crt, crt[0:1024])
+			},
+			certsReloadErrorAssert: require.Error,
+			certsAssert: func(t *testing.T, before []tls.Certificate, after []tls.Certificate) {
+				// No cert has been updated.
+				require.Len(t, before, 2)
+				require.Len(t, after, 2)
+				require.Equal(t, before[0], after[0])
+				require.Equal(t, before[1], after[1])
+			},
+		},
+		{
+			desc: "c1 and c2 certs do not change if one of them is corrupted",
+			certsUpdate: func(t *testing.T, certs *certFiles) {
+				// Corrupt c1 cert key.
+				_, err := certs.c1Key.WriteAt([]byte{1, 2, 3, 4, 5, 6, 7, 8}, 0)
+				require.NoError(t, err)
+				err = certs.c1Key.Sync()
+				require.NoError(t, err)
+			},
+			certsReloadErrorAssert: require.Error,
+			certsAssert: func(t *testing.T, before []tls.Certificate, after []tls.Certificate) {
+				// No cert has been updated.
+				require.Len(t, before, 2)
+				require.Len(t, after, 2)
+				require.Equal(t, before[0], after[0])
+				require.Equal(t, before[1], after[1])
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			// Create empty certs and ensure they get cleaned up.
+			certs := newCerts(t)
+			defer certs.cleanup(t)
```

Comment:
```suggestion
			t.Cleanup(certs.cleaup)
```

You'll also have to change the signature of `certs.cleanup` to no longer take a `testing.T`. 

Though, on second though, I would get rid of the cleanup method and have `newCerts` register the appropriate cleanup action (see below comments)